File: MyungJi/240101_week/programmers_154540_무인도여행.cpp
#include <bits/stdc++.h>

using namespace std;

int visited[104][104];
int dy[4]= {-1, 0, 1, 0};
int dx[4]= {0, 1, 0, -1};

int dfs(vector<string> &maps, int y, int x){
    visited[y][x]=1;
    int sum = (maps[y][x] - '0');
    for(int i=0;i<4;i++){
        int ny = y+ dy[i];
        int nx = x + dx[i];
        if (ny<0 || nx<0 || ny>=maps.size() || nx>=maps[0].size()) continue;
        if (visited[ny][nx]) continue;
        if (maps[ny][nx]!='X'){
            sum += dfs(maps, ny, nx);   
        }
    }  
    return sum;
}


vector<int> solution(vector<string> maps) {
    vector<int> answer;
    //1. dfs 수행
    for (int i=0;i<maps.size();i++){
        for(int j=0;j<maps[i].size();j++){
            if (maps[i][j]!='X' && visited[i][j]==0){
                answer.push_back(dfs(maps, i, j));
            }              
        }
    }
    if (answer.size()==0) return {-1};
    //2. 배열 정렬
    sort(answer.begin(), answer.end());
    return answer;
}

File: SooYeon/240101_week/Programmers_154540_무인도 여행.js
function solution(maps) {
    const answer = [];
    
    let n = maps.length;
    let m = maps[0].length;
    
    let dy = [-1, 0, 1, 0];
    let dx = [0, 1, 0, -1];
    
    let visited = Array.from(Array(n),() => new Array(m).fill(false));
    
    const dfs = (y, x) => {
        visited[y][x] = true;
        let foodSum = Number(maps[y][x]);
        
        for(let i = 0; i < 4; i++){
            let ny = y + dy[i];
            let nx = x + dx[i];
            
            if(ny >= n || ny < 0 || nx >= m || nx < 0) continue;
            if(maps[ny][nx] === 'X' || visited[ny][nx]) continue;
            
            foodSum += dfs(ny,nx);
        }
        
        return foodSum;
    }
    
    for(let i = 0; i < n; i++){
        for(let j = 0; j < m; j++){
            if(maps[i][j] !== 'X' && !visited[i][j]){
                answer.push(dfs(i, j));
            }
        }
    }
    if(answer.length === 0){
        return [-1];
    } 
    else {
        answer.sort((a, b) => a - b); // 오름차순 정렬
        return answer; 
    }

}
